aten/src/ATen/native/SpectralOps.cpp

```diff
@@ -5,33 +5,152 @@
 #endif
 
 #include "ATen/ATen.h"
+#include "ATen/Config.h"
 #include "ATen/NativeFunctions.h"
-#include "ATen/WrapDimUtils.h"
-#include "ATen/ExpandUtils.h"
+#include "ATen/native/SpectralOpsUtils.h"
 
 #include <algorithm>
-#include <functional>
-#include <numeric>
 #include <vector>
-
+#include <cmath>
 
 namespace at { namespace native {
 
-// Since real-to-complex satisfy the Hermitian symmetry, i.e.,
-// X[m, \omega] = X[m, N - \omega]*. We return only the first floor(N / 2) + 1
-// values by default. This is also the assumption in libraries including cuFFT.
-static inline int64_t infer_ft_complex_length(int64_t real_length) {
-  return (real_length >> 1) + 1;
+static inline Tensor _fft(const Tensor &self, const int64_t signal_ndim,
+           const bool complex_input, const bool complex_output,
+           const bool inverse, IntList signal_sizes, const bool normalized,
+           const bool onesided) {
+
+  if (signal_ndim < 1 || signal_ndim > 3) {
+    std::ostringstream ss;
+    ss << "Expected signal_ndim to be 1, 2, or 3, but got signal_ndim="
+       << signal_ndim;
+    throw std::runtime_error(ss.str());
+  }
+  if (!at::isFloatingType(self.type().scalarType())) {
+    std::ostringstream ss;
+    ss << "Expected an input tensor of floating types, but got input "
+       << self.type() << self.sizes();
+    throw std::runtime_error(ss.str());
+  }
+
+  auto signal_tensor_ndim = signal_ndim + (int) complex_input;  // add complex dim
```

Comment:
More idiomatic C++: `static_cast<int>(complex_input)`

Reply:
Also, you're using a contiguous inner dimension to simulate complex numbers. Is the contiguity of the inner dimension being checked somewhere? I couldn't find it.